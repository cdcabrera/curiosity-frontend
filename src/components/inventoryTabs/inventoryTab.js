/**
 * @memberof InventoryTabs
 * @module InventoryTab
 */

/**
 * A tab pass-through component for passing props to InventoryTabs.
 *
 * @param {object} props
 * @param {boolean} props.active
 * @param {React.ReactNode} props.children
 * @param {string} props.title
 * @returns {React.ReactNode}
 */
// eslint-disable-next-line no-unused-vars
const InventoryTab = ({ active, children, title }) => children;

export { InventoryTab as default, InventoryTab };
