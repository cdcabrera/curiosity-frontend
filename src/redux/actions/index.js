import { platformActions } from './platformActions';

const actions = {
  platform: platformActions
};

const reduxActions = { ...actions };

export { reduxActions as default, reduxActions, platformActions };
