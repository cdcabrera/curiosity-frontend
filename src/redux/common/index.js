import { reduxHelpers } from './reduxHelpers';

/**
 * @memberof Redux State
 * @module Helpers
 * @property {module} ReduxHelpers
 */

export { reduxHelpers as default, reduxHelpers };
