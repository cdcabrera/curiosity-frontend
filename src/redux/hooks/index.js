import { reactReduxHooks } from './useReactRedux';

/**
 * @memberof Redux State
 * @module Hooks
 * @property {module} UseReactRedux
 */

/**
 * Store hooks
 *
 * @type {{reactReduxHooks: {shallowEqual: Function, useDispatch: Function, useSelector: Function, useSelectors: Function,
 *     useSelectorsResponse: Function, useSelectorsAllSettledResponse: Function, useSelectorsAnyResponse: Function,
 *     useSelectorsRaceResponse: Function}}}
 */
const storeHooks = {
  reactRedux: reactReduxHooks
};

export { storeHooks as default, storeHooks, reactReduxHooks };
