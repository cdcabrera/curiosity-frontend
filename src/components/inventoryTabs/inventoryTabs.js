import React from 'react';
import PropTypes from 'prop-types';
import { Title } from '@patternfly/react-core';
import { connect, reduxTypes, store } from '../../redux';
import { Tabs } from '../tabs/tabs';
import { helpers } from '../../common';
import { translate } from '../i18n/i18n';

/**
 * ToDo: Revisit using tabs attribute vs children
 * Our initial tab implementation used children. This come across as backtracking, however upon
 * review with a ProductView POC using an array/list based on a product config was easier to
 * handle programmatically. Related to not every product having access to all tabs, or certain
 * products having a different default tab.
 */
/**
 * A system inventory tabs component.
 *
 * @augments React.Component
 * @fires onTab
 */
class InventoryTabs extends React.Component {
  /**
   * On tab update state.
   *
   * @event onTab
   * @param {object} params
   * @param {string} params.index tab index
   */
  onTab = ({ index }) => {
    const { productId } = this.props;

    store.dispatch({
      type: reduxTypes.inventory.SET_INVENTORY_TAB,
      tabs: {
        [productId]: index
      }
    });
  };

  /**
   * Render inventory tabs.
   *
   * @returns {Node}
   */
  render() {
    const { activeTab, isDisabled, tabs, t } = this.props;

    if (isDisabled) {
      return null;
    }

    return (
      <React.Fragment>
        <Title headingLevel="h2" className="sr-only">
          {t('curiosity-inventory.tabHeading')}
        </Title>
        <Tabs activeTab={activeTab} onTab={this.onTab} tabs={tabs} />
      </React.Fragment>
    );
  }
}

/**
 * Prop types.
 *
 * @type {{productId: string, tabs: Array, isDisabled: boolean, activeTab: number}}
 */
InventoryTabs.propTypes = {
  activeTab: PropTypes.number,
  isDisabled: PropTypes.bool,
  productId: PropTypes.string.isRequired,
  tabs: Tabs.propTypes.tabs.isRequired,
  t: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{isDisabled: boolean, activeTab: number}}
 */
InventoryTabs.defaultProps = {
  activeTab: 0,
  isDisabled: helpers.UI_DISABLED_TABLE,
  t: translate
};

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const mapStateToProps = ({ inventory }, { productId }) => ({ activeTab: inventory.tabs?.[productId] });

const ConnectedInventoryTabs = connect(mapStateToProps)(InventoryTabs);

export { ConnectedInventoryTabs as default, ConnectedInventoryTabs, InventoryTabs };
