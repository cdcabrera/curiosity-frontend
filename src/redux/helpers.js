import _camelCase from 'lodash/camelCase';
import _snakeCase from 'lodash/snakeCase';
import { helpers } from '../common';

/**
 * @memberof Helpers
 * @module ReduxHelpers
 */

/**
 * Set an API query based on specific API "acceptable values" schema.
 *
 * @param {object} values
 * @param {object} schema
 * @param {*} [initialValue]
 * @returns {object}
 */
const setApiQuery = (values, schema, initialValue) => {
  const generated = {};
  const schemaArr = (schema && Object.values(schema)) || [];

  schemaArr.forEach(value => {
    if (initialValue === undefined) {
      if (value in values) {
        generated[value] = values?.[value];
      }
    } else {
      generated[value] = values?.[value] || initialValue;
    }
  });

  return generated;
};

// ToDo: research applying a maintained schema map/normalizer such as, normalizr
/**
 * Apply a set of schemas using either an array of objects in the
 * form of [{ madeUpKey: 'some_api_key' }], or an array of arrays
 * in the form of [['some_api_key','another_api_key']]
 *
 * @param {Array} schemas
 * @param {*} [initialValue]
 * @returns {Array}
 */
const setResponseSchemas = (schemas = [], initialValue) =>
  schemas.map(schema => {
    const generated = {};
    const arr = (Array.isArray(schema) && schema) || Object.values(schema);

    arr.forEach(value => {
      generated[value] = initialValue;
    });

    return generated;
  });

/**
 * Normalize an API response.
 *
 * @param {*} responses
 * @param {object} responses.response
 * @param {object} responses.response.schema
 * @param {Array|object} responses.response.data
 * @param {string} responses.response.keyCase
 * @param {Function} responses.response.customResponseEntry
 * @param {Function} responses.response.customResponseValue
 * @param {string} responses.response.keyPrefix
 * @returns {Array}
 */
const setNormalizedResponse = (...responses) => {
  const parsedResponses = [];

  responses.forEach(
    ({ schema = {}, data, customResponseEntry, customResponseValue, keyPrefix: prefix, keyCase = 'camel' }) => {
      const isArray = Array.isArray(data);
      const updatedData = (isArray && data) || (data && [data]) || [];
      const [generatedSchema = {}] = setResponseSchemas([schema]);
      const parsedResponse = [];

      updatedData.forEach((value, index) => {
        const generateReflectedData = ({
          dataObj,
          keyPrefix = '',
          keyCaseType,
          customEntry,
          customValue = null,
          update = helpers.noop
        }) => {
          let updatedDataObj = {};

          Object.entries(dataObj).forEach(([dataObjKey, dataObjValue]) => {
            let casedDataObjKey;

            switch (keyCaseType) {
              case 'camel':
                casedDataObjKey = _camelCase(`${keyPrefix} ${dataObjKey}`).trim();
                break;
              case 'snake':
                casedDataObjKey = _snakeCase(`${keyPrefix} ${dataObjKey}`).trim();
                break;
              case 'default':
              default:
                casedDataObjKey = `${dataObjKey}`.trim();
                break;
            }

            let val = dataObjValue;

            if (typeof val === 'number') {
              val = (Number.isInteger(val) && Number.parseInt(val, 10)) || Number.parseFloat(val) || val;
            }

            if (typeof customValue === 'function') {
              updatedDataObj[casedDataObjKey] = customValue({ data: dataObj, key: dataObjKey, value: val, index });
            } else {
              updatedDataObj[casedDataObjKey] = val;
            }
          });

          if (typeof customEntry === 'function') {
            updatedDataObj = customEntry(updatedDataObj, index);
          }

          update(updatedDataObj);
        };

        generateReflectedData({
          keyPrefix: prefix,
          dataObj: { ...generatedSchema, ...value },
          keyCaseType: keyCase,
          customEntry: customResponseEntry,
          customValue: customResponseValue,
          update: generatedData => parsedResponse.push(generatedData)
        });
      });

      parsedResponses.push(parsedResponse);
    }
  );

  return parsedResponses;
};

const reduxHelpers = {
  setApiQuery,
  setResponseSchemas,
  setNormalizedResponse
};

export { reduxHelpers as default, reduxHelpers };
