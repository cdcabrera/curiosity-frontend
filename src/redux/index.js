import { connect, useSelector as UseSelector } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { store } from './store';
import { reduxActions } from './actions';
import { reduxHelpers, apiQueries } from './common';
import { reduxReducers } from './reducers';
import { reduxSelectors } from './selectors';
import { reduxTypes } from './types';
import { helpers } from '../common';

const connectRouter = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(connect(mapStateToProps, mapDispatchToProps)(component));

const useSelector = (selector, value = null, options = {}) =>
  (helpers.TEST_MODE && value) || UseSelector(selector, options.equality) || value;

export {
  apiQueries,
  connect,
  connectRouter,
  reduxActions,
  reduxHelpers,
  reduxReducers,
  reduxSelectors,
  reduxTypes,
  store,
  useSelector
};
