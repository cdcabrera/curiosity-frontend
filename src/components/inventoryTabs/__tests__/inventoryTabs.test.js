import React from 'react';
import { mount, shallow } from 'enzyme';
import { InventoryTabs } from '../inventoryTabs';
import { store } from '../../../redux';

describe('InventoryList Component', () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.spyOn(store, 'dispatch').mockImplementation((type, data) => ({ type, data }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a non-connected component', () => {
    const props = {
      productId: 'lorem',
      tabs: [
        { title: 'lorem', content: 'ipsum' },
        { title: 'sit', content: 'amet' }
      ]
    };

    const component = shallow(<InventoryTabs {...props} />);
    expect(component).toMatchSnapshot('non-connected');
  });

  it('should return an empty render when disabled', () => {
    const props = {
      productId: 'lorem',
      tabs: [
        { title: 'lorem', content: 'ipsum' },
        { title: 'sit', content: 'amet' }
      ],
      isDisabled: true
    };
    const component = shallow(<InventoryTabs {...props} />);

    expect(component).toMatchSnapshot('disabled component');
  });

  it('should handle updating tabs for redux state', () => {
    const props = {
      productId: 'lorem',
      tabs: [
        { title: 'lorem', content: 'ipsum' },
        { title: 'sit', content: 'amet' }
      ]
    };
    const component = mount(<InventoryTabs {...props} />);
    const componentInstance = component.instance();

    componentInstance.onTab({ index: 1 });
    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch filter');
  });
});
