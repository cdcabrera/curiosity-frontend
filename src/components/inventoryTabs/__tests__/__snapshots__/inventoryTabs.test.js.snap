// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryList Component should handle updating tabs for redux state: dispatch filter 1`] = `
Array [
  Array [
    Object {
      "tabs": Object {
        "lorem": 1,
      },
      "type": "SET_INVENTORY_TAB",
    },
  ],
]
`;

exports[`InventoryList Component should render a non-connected component: non-connected 1`] = `
<Tabs
  activeTab={0}
  className=""
  defaultActiveTab={0}
  hasOverflowScroll={false}
  onTab={[Function]}
  tabs={
    Array [
      Object {
        "content": "ipsum",
        "title": "lorem",
      },
      Object {
        "content": "amet",
        "title": "sit",
      },
    ]
  }
/>
`;

exports[`InventoryList Component should return an empty render when disabled: disabled component 1`] = `""`;
